src/charm.py

```diff
@@ -1,233 +1,294 @@
+
+        try:
+            self._container.restart(WORKLOAD_CONTAINER)
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus(
+                "Failed to restart the container, please consult the logs"
+            )
+            return
 
     @log_event_handler(logger)
-    def _on_pebble_ready(self, event):
+    def _on_pebble_ready(self, event: PebbleReadyEvent) -> None:
         """Handle pebble ready event.
 
+            - open application port
+            - handle status and update config
+            - set workload version.
+
         Args:
-            event: The event triggered when the relation changed.
+            event: The event triggered when Pebble is ready for a workload.
         """
-        self._update(event)
+        self.model.unit.open_port(protocol="tcp", port=OAUTH2_PROXY_API_PORT)
+        self._handle_status_update_config(event)
+        self.unit.set_workload_version(OAUTH2_PROXY_VERSION)
 
     @log_event_handler(logger)
-    def _on_config_changed(self, event):
-        """Handle configuration changes.
+    def _on_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
+        """Handle ingress ready event and update oauth relation data.
 
         Args:
-            event: The event triggered when the relation changed.
+            event: The event triggered when IngressPerApp is ready.
         """
-        self.unit.status = WaitingStatus("configuring application")
-        self._update(event)
+        if self.unit.is_leader():
+            logger.info(f"This app's ingress URL: {event.url}")
+            self.oauth.update_client_config(client_config=self._oauth_client_config)
 
     @log_event_handler(logger)
-    def _on_update_status(self, event):
-        """Handle `update-status` events.
+    def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
+        """Handle ingress revoked event and update oauth relation data.
 
         Args:
-            event: The `update-status` event triggered at intervals.
+            event: The event triggered when IngressPerAppRevoked is emitted.
         """
-        try:
-            self._validate()
-        except ValueError:
-            return
-
-        container = self.unit.get_container(self.name)
-        valid_pebble_plan = self._validate_pebble_plan(container)
-        if not valid_pebble_plan:
-            self._update(event)
-            return
-
-        check = container.get_check("up")
-        if check.status != CheckStatus.UP:
-            self.unit.status = MaintenanceStatus("Status check: DOWN")
-            return
-
-        self.unit.set_workload_version(OAUTH2_PROXY_VERSION)
-        self.unit.status = ActiveStatus()
+        if self.unit.is_leader():
+            logger.info("This app no longer has ingress")
+            self.oauth.update_client_config(client_config=self._oauth_client_config)
```

Comment:
Don't we need to reconfigure all the pods if this happens? 
```suggestion
        if self.unit.is_leader():
            logger.info("This app no longer has ingress")
            self.oauth.update_client_config(client_config=self._oauth_client_config)
        self._handle_status_update_config(event)
```

Reply:
Right now there is nothing to be reconfigured in the pods when ingress departs, as oauth2 proxy config doesn't depend on it (yet)